### PR TITLE
Add threadlocal bound context manager

### PR DESCRIPTION
**Pull request by fernandobrito**
Adds a context manager/decorator similar to `tmp_bind`, but using the thread-local context instead.

Follow-up from https://github.com/hynek/structlog/issues/218.

I have some questions before I start with documentation and improve the tests.
I'm also new to `structlog`, so let me know if I'm missing something obvious.

By the way, great PR checklist, contributing guide and setup with `pre-commit` 🙌 .

## Clean up

When exiting the context manager, how should we handle unbiding?

A) We could just `unbind_threadlocal` all keys that were passed as parameters to the context manager (like this version of the PR), but if there was already an existing key in the context before the use of `bound` and `bound` overwrites it with another value, in this approach the teardown step will actually remove whatever value was there before. See sample test `test_cleanup_conflict` in this PR.

B) `tmp_bind` creates a copy of the context and "rewinds" it when tearing it down. If we use this approach here, we will also revert any other independent use of `bind_threadlocal` or `unbind_threadlocal` that happened inside the managed context. See sample test `test_cleanup_conflict2` in this PR. Personally, I prefer the approach above since it seems less likely to affect users than this one.

What are your thoughts on it? Keep it simple, use approach A and make it clear in the documentation that conflicting keys will be removed after the context manager is done? Or go nuts and try an option C where we would check what keys `bound` is overwriting, then we keep only a copy of those (and not of the whole context, like approach B), and on teardown we first `unbind` all keys passed to `bound` and then merge this copy back to the context? Is it worth complicating it like that?

Let me know if this all makes sense or if I should try to rephrase my thoughts 😁

# Pull Request Check List

This is just a friendly reminder about the most common mistakes.  Please make sure that you tick all boxes.  But please read our [contribution guide](https://www.structlog.org/en/latest/contributing.html) at least once, it will save you unnecessary review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing left to do.

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.
    - [x] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).  Find the appropriate next version in our [``__init__.py``](https://github.com/hynek/structlog/blob/main/src/structlog/__init__.py) file.
- [x] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x] Changes (and possible deprecations) are documented in the [changelog](https://github.com/hynek/structlog/blob/main/CHANGELOG.rst).

If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing!

### CODE REVIEW

**Comment by hynek** [comment]
I think as a user, I'd expect my old key to be restored and nothing else.

IOW:

```python
threadlocal_bind(x=1, y=10)
with threadlocal_bound(x=2):  # please change to this name :)
    log.info("foo") # logs x=2, y=10

 log.info("foo") # logs x=1, y=10
```

Your nuclear option would be more of a snapshot feature than a temporary bind. 😅

---

I guess you'll have to save which keys you need to delete and which you need to restore before binding and then do it in finally.

**Comment by fernandobrito** [comment]
Great, thanks for taking a look! I agree with you, and I will address it when time allows (probably only in the next weekend).

**Comment by hynek** [comment]
Cool! Since we've had a minor regression in #373, I'd like to push a new release sometime next week. Would be great if we could ship this along with it!

File: docs/api.rst [hunk 1]
```diff
@@ -119,6 +119,10 @@ Modern Approach
 
 .. autofunction:: bind_threadlocal
 
+.. autofunction:: unbind_threadlocal
```
Comment (fernandobrito) [inline docs/api.rst:122]: `unbind_threadlocal` was also missing from here.

**Comment by fernandobrito** [comment]
@hynek AwesomeI just did some small changes and improved the overwriting behavior/restoring process.

Regarding the documentation for the end-users, I'm not sure where to add it.
Currently, `tmp_bind` is actually inside the `Wrapped Dicts` section of `thread-local`: https://www.structlog.org/en/stable/thread-local.html#wrapped-dicts. Should I move `tmp_bind` to a new section and add `bound_threadlocal` close to it, or what would you suggest?

File: CHANGELOG.rst [hunk 2]
```diff
@@ -26,6 +26,9 @@ Changes:
 
 - Fixed import when running in optimized mode (``PYTHONOPTIMIZE=2``).
   `#373 <https://github.com/hynek/structlog/pull/373>`_
+- Added ``structlog.threadlocal.bound_threadlocal`` as a thread-local equivalent of ``structlog.threadlocal.tmp_bind``
+  for temporarily binding keys to the global context through a decorator/context manager.
```
Comment (hynek) [inline CHANGELOG.rst:30]: Let's not "advertise" the old APIs. Also: semantic newlines please. :)

```suggestion
- Added the ``structlog.threadlocal.bound_threadlocal`` decorator/context manager to temporarily bind key/value pairs to a thread-local context.
```

File: src/structlog/threadlocal.py [hunk 3]
```diff
@@ -93,6 +93,9 @@ def tmp_bind(
 ) -> Generator[TLLogger, None, None]:
     """
     Bind *tmp_values* to *logger* & memorize current state. Rewind afterwards.
+
+    Use :func:`~structlog.threadlocal.bound_threadlocal` instead when you want
+    the temporary bindings to be applied to the global context (thread-local).
```
Comment (hynek) [inline src/structlog/threadlocal.py:98]:
```suggestion
    Only works with `structlog.threadlocal.wrap_dict`-based contexts.
    Use :func:`~structlog.threadlocal.bound_threadlocal` for new code.
```

File: src/structlog/threadlocal.py [hunk 4]
```diff
@@ -255,6 +258,28 @@ def unbind_threadlocal(*keys: str) -> None:
...
 
 
+@contextlib.contextmanager
+def bound_threadlocal(**kw: Any) -> None:
+    """
+    Bind *kw* to the current thread-local context. Unbind *kw* afterwards &
+    restore overwritten keys. Can be used as a context manager or decorator.
+
+    Use this instead of :func:`~structlog.threadlocal.tmp_bind` when you want
+    the temporary bindings to be applied to the global context (thread-local).
```
Comment (hynek) [inline src/structlog/threadlocal.py:268]: Let's not point to `tmp_bind`, the new APIs stand for themselves:

```suggestion
    Bind *kw* to the current thread-local context. Unbind or restore *kw*
    afterwards. Do **not** affect other keys.

    Can be used as a context manager or decorator.
```

**Review by hynek — CHANGES_REQUESTED** [review]
Looking good!

Just please:

- add an example to tests/typing_examples.py
- add an example to the big example box in thread-local.rst, line 46pp

**Comment by hynek** [comment]
> Regarding the documentation for the end-users, I'm not sure where to add it. Currently, `tmp_bind` is actually inside the `Wrapped Dicts` section of `thread-local`: https://www.structlog.org/en/stable/thread-local.html#wrapped-dicts. Should I move `tmp_bind` to a new section and add `bound_threadlocal` close to it, or what would you suggest?

No, please leave `tmp_bind` alone, I want people to forget about it. 😅

Just add it to the big doctest in line 46 et pp. :)

File: CHANGELOG.rst [hunk 5]
```diff
@@ -26,6 +26,9 @@ Changes:
...
+ `#371 <https://github.com/hynek/structlog/pull/371>`_
```
Comment (hynek) [inline CHANGELOG.rst:31]:
```suggestion
- Added the ``structlog.threadlocal.bound_threadlocal`` decorator/context manager to temporarily bind key/value pairs to a thread-local context.
  `#371 <https://github.com/hynek/structlog/pull/371>`_
```

**Comment by hynek** [comment]
Looks like the types need some love: https://github.com/hynek/structlog/runs/4320516792?check_suite_focus=true#step:5:17

Is that something you're comfortable to tackle?

**Comment by fernandobrito** [comment]
> Looks like the types need some love: https://github.com/hynek/structlog/runs/4320516792?check_suite_focus=true#step:5:17
> 
> Is that something you're comfortable to tackle?

Yes, I think so, especially because there are other examples in the code already and `mypy` is already pointing me somewhere, but since tonight I will be busy, I will only be able to work on it tomorrow evening.

**Comment by hynek** [comment]
I think I've got it – thanks!

**Comment by fernandobrito** [comment]
Awesome, thanks! 🙌

**Pull request merged**